pre-piksi-multi-launch-updates.md

```diff
@@ -0,0 +1,222 @@
+  - Bits 3-7: Reserved
+- Have observations be sent out at all times, use flags above to indicate validity
+- Add measured doppler field as a Q16.8 layout
+  - 16 bit signed part (-32768 to 32767), 1/256 hertz residual
+- Deprecate current Tracking State Message as it is no longer needed when we send out observations for all tracked signals with the flags
+
+
+### 3. sender_id = device_uuid & 0xFFFF
+
+We need to decide where sender_id will come from on Piksi Multi. In deciding
+this, we need to keep in mind the following goals:
+- sender_id *does not* need to be globally unique amongst all Piksi Multi's
+- sender_id *does* need to be unique amongst the Piksi Multi's on a customer's
+local network of devices (for instance, point to multi-point ISM radio network,
+TCP/IP network, etc)
+
+- Decision is to use the lower 16 bits of the device UUID for serial numbers
+- We store the serial number to UUID map in our bringup database
+- We work with OPS to put the lower 16 bits of UUID on the physical hardware
+
+### 4. Remove request / response messages where possible
+
+The following request response messages are not used in Piksi Multi and can be
+deprecated:
+
+- `MSG_BOOTLOADER_HANDSHAKE_REQ`
+- `MSG_BOOTLOADER_HANDSHAKE_RESP`
+- `MSG_FLASH_READ_REQ`
+- `MSG_FLASH_READ_RESP`
+- `MSG_FLASH_PROGRAM`
+- `MSG_FLASH_ERASE`
+- `MSG_FLASH_DONE`
+- `MSG_STM_FLASH_LOCK_SECTOR`
+- `MSG_STM_FLASH_UNLOCK_SECTOR`
+- `MSG_M25_FLASH_WRITE_STATUS`a
+- `STM_UNIQUE_ID`
+
+The following messages could be part of a device ID message that is
+periodically sent:
+
+- `MSG_NAP_DEVICE_DNA_REQ`
+- `MSG_NAP_DEVICE_DNA_RESP`
+- `MSG_STM_UNIQUE_ID_REQ`
+- `MSG_STM_UNIQUE_ID_RESP`
+
+Some items will be exposed through read only settings interface:
+
+- `Device DNA`
+- `Buildroot version` 
+-  etc
+
+
+### 5. Replace obs forwarding with general purpose forwarding message
+
+Current there is a convention in the protocol that if the sender_id of a
+MSG_OBS is 0, it means it is a forwarded observation from another device. We
+propose to replace this with a general purpose forwarding message that wraps
+the forwarded message, discarding the preamble and checksum of the forwarded
+message. This way, the sender_id of the forwarded message will be preserved and
+will remove the need for special handling of sender_id = 0 on obs messages.
+
+The forwarded message will have three fields
+
+- u8 source ID :  Enum for different sources (uart0, uart1, ethernet, cell modem)
+- u8 protocol ID : enum for different protocols (SBP, RTCMv3)
+- u8[variable length] forwarded msg stream:
+
+### 6. Add UTC time message
+
+Users want UTC time rather than GPS time.  message format is below.  When Leap seconds is hardcoded from library (and not from sky) flags is 1.  When we get leap seconds from sky, we go to flags 2
+  - u8     flags
+  - u32     TOW [ms]
+  - u16    year
+  - u8     month
+  - u8     day
+  - u8    hours
+  - u8     minutes
+  - u8    seconds
+  - s32     ns    (nanoseconds)
```

Comment:
I think the calculated nanosecond remainder on the receiver can actually be negative.